src/Illuminate/Filesystem/Filesystem.php

```diff
@@ -122,6 +123,40 @@ public function put($path, $contents, $lock = false)
         return file_put_contents($path, $contents, $lock ? LOCK_EX : 0);
     }
 
+    /**
+     * Write the contents of a file, replacing it atomically if it already exists.
+     *
+     * @param string $path
+     * @param string $content
+     * @throws Exception
+     */
+    public function replace($path, $content)
+    {
+        // Just in case path already exists and is a symlink, we want to make sure we get the real path.
+        clearstatcache(true, $path);
+        $realPath = realpath($path);
+        if ($realPath) {
+            $path = $realPath;
+        }
+
+        $dirName = dirname($path);
+        if (! is_writable($dirName)) {
+            throw new Exception("Replacing $path requires it's parent directory to be writable.");
+        }
+
+        // Write out the contents to a temp file, so we can rename the file atomically.
+        $tempPath = tempnam($dirName, basename($path));
+        $this->put($tempPath, $content);
+
+        if (file_exists($path)) {
+            // Copy over the permissions and owner from the original file.
+            chmod($tempPath, (int) base_convert(substr((string) decoct(fileperms($path)), -3), 8, 10));
```

Comment:
Does this `fileperms`magic work on windows?

Reply:
Good point. The PHP docs on `fileperms()` doesn't mention Windows. I'll have to test this on a Windows machine.

If it doesn't, I'll just remove this line. That will make it function a little more different from `put()` in the corner case where the permissions of the existing file don't match those of the file it is replaced with. But for the places where we would want to use it it's still appropriate. `replace()` isn't a perfect substitute for `put()` to begin with.

Perhaps I am just overthinking things and I should remove the `chmod()` and `chgrp()` altogether and just make it clear `replace()` will also replace the ownership and permissions of the file.

Reply:
So I did what I was considering; I removed the `chmod()` and `chgrp()`. Also mentioned what happens to file permissions in the PHPDoc. This also solves any potential Windows incompatibilities.